bitcoin/src/psbt/mod.rs

```diff
@@ -385,6 +397,79 @@ impl Psbt {
         Ok(used)
     }
 
+    /// Attempts to create all signatures required by this PSBT's `tap_key_origins` field, adding
+    /// them to `tap_key_sig` or `tap_script_sigs`.
+    ///
+    /// # Returns
+    ///
+    /// - Ok: A list of the public keys used in signing.
+    /// - Err: Error encountered trying to calculate the sighash AND we had the signing key.
+    fn bip32_sign_schnorr<C, K, T>(
+        &mut self,
+        k: &K,
+        input_index: usize,
+        cache: &mut SighashCache<T>,
+        secp: &Secp256k1<C>,
+    ) -> Result<Vec<PublicKey>, SignError>
+        where
+            C: Signing,
+            T: Borrow<Transaction>,
+            K: GetKey,
+    {
+
+        let mut input = self.inputs[input_index].clone();
```

Comment:
As the algo is written the clone is necessary because we mutate `self.input` while iterating `tap_key_origins`.

Reply:
The reason for doing this is as @tcharding explained.

Reply:
@tcharding can you point at the conflicting lines? I don't see where the whole `Input` is mutably borrowed while iterated. Note that iterating `tap_key_origins` while mutating other parts of `Input` is fine.

Reply:
I apologize for any confusion caused earlier. Allow me to provide a detailed explanation.

If we avoid using clone and instead utilize a mutable reference, when the code reaches sighash_taproot, it will result in an immutable borrow of self (Psbt). Even if we modify the sighash_taproot interface to &mut self, the compiler will still check this line, and the algorithm later relies on the usage of the first mutable reference, leading to a compilation failure.

In this algorithm, using clone is indeed the solution to this problem. If there are any alternative approaches, please let me know, and I will correct it accordingly.

Reply:
>  and instead utilize a mutable reference

I'm not suggesting to use mutable reference either. I'm suggesting to index it on every access (and maybe some of those can go through reference). I think the check will be removed by the optimizer so there should be no perf penalty.

I may look into it when I have a bit of time.

Reply:
I tried replacing `input` with `self.inputs[input_index]` as you suggested, but it doesn't seem to work because when compiling `self.inputs[input_index] = Some(signature)`, it requires a mutable value. If we use a mutable reference, it brings us back to the same issue mentioned earlier. Given the current algorithm, it seems that the current approach is relatively appropriate. If there are better solutions in the future, we can modify it accordingly. For now, I will go ahead.

Reply:
`self` is mutable anyway, so I don't see a problem. But sure, this is not important and if you don't see an easy solution feel free to leave it as-is.

Reply:
Regarding this portion of the code, I will keep it as it is for now. However, I need to provide some additional information about this issue. Since self is mutable, the code `self.inputs[input_index].tap_key_sig = Some(signature)` is indeed valid, as you mentioned. However, the use of `self.inputs[input_index].tap_key_origins.iter()` along with `self.inputs[input_index].tap_key_sig = Some(signature)` will trigger the "second mutable borrow" problem.